### DISCUSSION TITLE
Codegen: How to get a list and iterate over the items?

### DISCUSSION THREAD

**Discussion by narcoticfresh** [activej/activej#320] [Q&A] [answered]
Hi guys..

I want to generate some block of code using activej-codegen that

* calls a function that returns a `List`
* iterates the members of that `List`, calling another function for each value (basically what `.map()` does)
* get a new list with the modified entries

i tried this

```java
Expression iter = let(
      // new map
      constructor(ArrayList.class),
      newList -> let(
        staticCall(ValueSplitter.class, "splitValue", fieldValue, value(field.getExplodeBy())),
        splittedList -> iterateList(
          splittedList,
          listMember -> call(newList, "add", getFieldConverterFunction(field.getType(), listMember))
        )
      )
    );
```

When i try to run this, being embedded in my function expression (but it did work), it crashes with ([full output is here](https://gist.github.com/narcoticfresh/94106c0c74102c9fe72ca316ebe5611e)):

```
java.lang.VerifyError: Inconsistent stackmap frames at branch target 846
Exception Details:
  Location:
    io/activej/codegen/app/targetloader/util/rowmapper/RowMapper_1.mapRow(Ljava/util/HashMap;)Lorg/bson/Document; @841: goto
```

Is this a correct approach? Or how would one do this? Sadly, documentation for codegen (i mostly [use the tests for reference](https://github.com/akumaburn/activej/blob/5ff0655d0fbfd76f20faa2d2e548037d1c4ceec8/core-codegen/src/test/java/io/activej/codegen/expression/ExpressionTest.java))  don't provide more information.

Also, there is no `iterateList` tests in the linked test class.
Generally I don't get how `sequence()` and the "local scope" with `let()` really works. 

Also my general question: How does one "debug"/"analyze" what exactly the problem here is?  What is the approach to find out where problem is?

Thanks for any help!

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @narcoticfresh 

I think your approach is correct. There might be subtle errors, that I can not verify having just that snippet.

I have created [an example](https://gist.github.com/eduard-vasinskyi/bc2ebdde5bb32f36e22d65928387ee09) that follows your snippet. It generates a class that has a `getMappedList()` method. The implementation of `getMappedList()` creates a new ArrayList. Then it calls the static method `ListHolder.getList()` to obtain a list. After that, it iterates over elements of the list and adds 100 to each. The result is added to the created ArrayList. Finally, the list is returned from the method.

You could modify that example to your code.

As for the `let` method, it creates a new local variable from an expression and then it applies some function to that variable. The resulting expression is the one that is returned from that function.

The `sequence` method executes a sequence of expressions and the resulting expression is the last in the sequence.

In my example, I want the `getMappedList` method to return a list that contains mapped elements. However, the `let` expression that creates the `splittedList` variable takes a function that calls the `iterateList` expression. The `iterateList` expression has a return type VOID. So, to return the `newList`, I used the `sequence` expression and passed `newList` as the second (and last) argument to the `sequence` expression.

If you want to debug the generated code, you can create a `DefiningClassLoader` specifying the debug output dir. This is `classLoaderWithDebug` in the example. If your code generates without an error, you can find a generated class inside the debug directory. Here is [the decompiled code](https://gist.github.com/eduard-vasinskyi/ebba54d4c159a688d167c8298056ded2) of the generated method.

**Reply by narcoticfresh** [reply]
@eduard-vasinskyi 
thank you so much for the example!

indeed the difference between

```java
Expression iter = let(
      // new map
      constructor(ArrayList.class),
      newList -> let(
        ...
      )
```

and (what you provided)

```java
Expression iter = let(
      // new map
      constructor(ArrayList.class),
      newList -> sequence(
        let(...)
      )
```

seems to have fixed it.. AFAIK i didn't change anything else..

> If you want to debug the generated code, you can create a DefiningClassLoader specifying the debug output dir. This is classLoaderWithDebug in the example. If your code generates without an error, you can find a generated class inside the debug directory. Here is [the decompiled code](https://gist.github.com/eduard-vasinskyi/ebba54d4c159a688d167c8298056ded2) of the generated method.

oh great, thanks.. 

are you aware of any possibilities to debug the error in case of a generation error? how to find where the problem is exactly? i guess that is not possible..

thanks again!
